File: Backend/src/main/java/tn/school/management/service/StudentService.java
package tn.school.management.service;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import tn.school.management.dto.StudentDto;
import tn.school.management.entity.Level;

public interface StudentService {
    
    StudentDto createStudent(StudentDto studentDto);
    
    StudentDto getStudentById(Long id);
    
    Page<StudentDto> getAllStudents(Pageable pageable);
    
    StudentDto updateStudent(Long id, StudentDto studentDto);
    
    void deleteStudent(Long id);
    
    Page<StudentDto> searchStudents(String username, Pageable pageable);
    
    Page<StudentDto> filterStudentsByLevel(Level level, Pageable pageable);
}

File: Frontend/src/app/services/student.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Page, Student, Level } from '../models/types';

@Injectable({
  providedIn: 'root'
})
export class StudentService {

  private apiUrl = 'http://localhost:8080/api/students';

  constructor(private http: HttpClient) { }

  getAllStudents(page: number = 0, size: number = 10): Observable<Page<Student>> {
    const params = new HttpParams()
      .set('page', page)
      .set('size', size);
    return this.http.get<Page<Student>>(this.apiUrl, { params });
  }

  getStudentById(id: number): Observable<Student> {
    return this.http.get<Student>(`${this.apiUrl}/${id}`);
  }

  createStudent(student: Student): Observable<Student> {
    return this.http.post<Student>(this.apiUrl, student);
  }

  updateStudent(id: number, student: Student): Observable<Student> {
    return this.http.put<Student>(`${this.apiUrl}/${id}`, student);
  }

  deleteStudent(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  searchStudents(username: string, page: number = 0, size: number = 10): Observable<Page<Student>> {
    const params = new HttpParams()
      .set('username', username)
      .set('page', page)
      .set('size', size);
    return this.http.get<Page<Student>>(`${this.apiUrl}/search`, { params });
  }

  filterByLevel(level: Level, page: number = 0, size: number = 10): Observable<Page<Student>> {
    const params = new HttpParams()
      .set('level', level)
      .set('page', page)
      .set('size', size);
    return this.http.get<Page<Student>>(`${this.apiUrl}/filter`, { params });
  }
}
